### PR TITLE
React 19 support

### PR DESCRIPTION
**Pull request by Mati365**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Add support to React 19 and create demos. Adjust context format to prevent race conditions in `CKEditor` component. 

### Additional Information

Collaboration samples update PR: https://github.com/cksource/ckeditor5-collaboration-samples/pull/146
Closes:
1. https://github.com/ckeditor/ckeditor5-react/issues/478 
2. https://github.com/ckeditor/ckeditor5-react/issues/312
3. https://github.com/ckeditor/ckeditor5-react/issues/485

May fix some of these bugs that seems to be related to Context:
1. https://github.com/ckeditor/ckeditor5-react/issues/476 (race condition)
4. https://github.com/ckeditor/ckeditor5-react/issues/294 (race condition)
5. https://github.com/ckeditor/ckeditor5-react/issues/409 (exported `ContextWatchdogContext` and `useCKEditorWatchdogContext` so now it is possible to implement own version of context used by `CKEditor`) 

### New features

1. Export context access hook `useCKEditorWatchdogContext`

```tsx
const Component = () => {
  const context = useCKEditorWatchdogContext();
  // ...
};
```

2. Export context `ContextWatchdogValue` typings and guards (`isContextWatchdogValue`, `isContextWatchdogValueWithStatus`).

```tsx
export type ContextWatchdogValue =
	| {
		status: 'initializing';
	}
	| {
		status: 'initialized';
		watchdog: ContextWatchdog;
	}
	| {
		status: 'error';
		error: ErrorDetails;
	};
```

3. Export context provider `ContextWatchdogContext`.

```tsx
const Component = () => {
   // ...
   return (
      <ContextWatchdogContext.Provider value={yourCustomContext}>
        ..
      </ContextWatchdogContext.Provider>
   );
};
```

❓ Exports above allow slightly better customization of contexts used by `CKEditor` component. Especially mentioned https://github.com/ckeditor/ckeditor5-react/issues/409 in this thread. 

#### New installation methods

All demos have been moved to new installation methods and use `ckeditor5` dependency.

#### `CKEditorContext` crashes

 `CKEditorContext` was using a class approach to define the component. However, with React 19, there have been changes in the calling order of lifecycle methods, which introduced race conditions in watchdog context creation that were fully reproducible in our demos. You can find an example [here](https://github.com/ckeditor/ckeditor5-react/issues/478#issuecomment-2120127393).
  
This error occurred because `CKEditorContext` created two instances of `CKEditorContext` in Strict Mode, resulting in the first watchdog instance being used in the children `CKEditor` component instead of the second one.
  
It was difficult to store additional information between the renders in a class component (constructor is called two times in strict mode), so I decided to refactor the component to a functional type. Unlike class components, functional components can hold some data between strict mode re-renders. Now, we store the last watchdog ID in a `useRef` and then check if the fully initialized watchdog ID matches the last requested watchdog ID. If it does, it is set to the state; otherwise, it is destroyed.

Child `CKEditorContext` components, such as `CKEditor` component, incorrectly implemented detection of context state. After mounting of child, it was unknown if child was placed in `CKEditorContext` or not because `CKEditorContext` default value (while being initialized in async lifecycle methods) was `null`. I added `initializing` state to check if child is being initialized in `CKEditorContext`.

**⚠️  Possible breaking changes** 

1. The `CKEditorContext` is no longer a class component, so these statements (which were highly error-prone and undocumented in our documentation) are no longer functional:

```tsx
const contextRef = useRef();

useEffect( () => {
   // Right now it is not possible to access `contextWatchdog` anymore and it will return `undefined`.
   const { contextWatchdog } = contextRef.current;
   
   // ...
}, [] );

<CKEditorContext
   ...
   ref={ contextRef }
/>
```

3. The `status` field has been added to the context type, indicating when the watchdog has been fully initialized. Current `Context` format:

```ts
export type ContextWatchdogValue =
	| {
		status: 'initializing';
	}
	| {
		status: 'initialized';
		watchdog: ContextWatchdog;
	}
	| {
		status: 'error';
		error: ErrorDetails;
	};
```

It simplifies deduction and prevents race condition in reusing watchdog logic inside components. Especially [here](https://github.com/ckeditor/ckeditor5-react/blob/81e323bbe0fdb594420a7ba859ec83767fd83dcd/src/ckeditor.tsx#L211). From now, the `CKEditor` component is waiting for fully initialization of watchdog to start bootstrapping editor.  

However, `CKEditorContext` (React Context) was not exported from the package, so it should not be used by external integrations. I would mark it as a potential breaking change because there is one additional re-render due to the change from initializing to initialized state set in the context.

#### `useMultiRootEditor` crashes

It was impossible to register new root in React 19 due to this [exception](https://github.com/ckeditor/ckeditor5-react/issues/478#issuecomment-2120002164). It happened because `EditorEditable` was rendered two times in strict mode and second execution was performed on the DOM element that was already assigned to editor.

### CODE REVIEW

**Comment by Mati365** [comment]
No regressions according to QA [testing results](https://cksource.slack.com/archives/C04VC9FK5LJ/p1716888581538939?thread_ts=1716378172.304639&cid=C04VC9FK5LJ).

**Comment by Mati365** [comment]
Do not rebase.

**Comment by Mati365** [comment]
Note after call: It'll be merged _after_ new installation methods.

**Comment by Mati365** [comment]
PR should now work with new installation methods

**Comment by Mati365** [comment]
@f1ames Can you take a look at this PR? I've heard that you worked with React before, it'd be nice to hear additional feedback about changes. Thanks!

File: src/useMultiRootEditor.tsx
Comment (DawidKossowski) [inline src/useMultiRootEditor.tsx:574]: I think you could use `editor.detachEditable` instead.

**Review by DawidKossowski — CHANGES_REQUESTED** [review]
Several additional things that I found:

1. Unit tests produce a lot of errors in the console:
Zrzut ekranu 2024-06-7 o 10 06 48
Is it the desired behavior?

2. I got this error once in the Context demo (React 19):
Zrzut ekranu 2024-06-7 o 10 41 58
However, I am no longer able to reproduce it. I believe there is still a possibility of a race condition. Perhaps we should invest more time in testing and debugging it.

3. The `isLayoutReady` prop in `CKEditorContext` works differently in the "classic" integration and multi-root integration. After providing `isLayoutReady={false}` in `demo-multiroot-react-19`, roots are rendered, and you can edit the content. However, when you pass the same prop in `demo-react-19`, the editor is not shown.

4. Generally, I understand it was a good opportunity to refactor some code, but adding new functionalities made the overall changes unclear. I'm not sure which changes were necessary in the context of supporting React 19 and which were made for refactoring purposes. I believe we could have done it in two steps. Additionally, I need to rethink the additional types, guards, and so on.

File: demo-react-19/src/ContextDemo.tsx
Comment (Mati365) [reply demo-react-19/src/ContextDemo.tsx:36]: It is still required, as it was mentioned in comment the react types are conflicting.

**Comment by Mati365** [comment]
@DawidKossowski 

> Unit tests produce a lot of errors in the console

It happens when CKEditor has roots that are not assigned to DOM while being destroyed. Happens frequently in `renderHook` test suite. I added fake elements before destroying editor to prevent this situation ([commit](https://github.com/ckeditor/ckeditor5-react/pull/480/commits/6b114d7a74887f6c658fcfd0b90ed4a513d61063)). Overall I think it's not super crucial and maybe useful in testing only. 
  
> I got this error once in the Context demo (React 19):

I cannot reproduce this race condition altough I believe that you are right. There can be race condition between synchronization of watchdog state and React state. Basically context might be destroyed and react state update (about reinitialization of such react context) is still pending while editor is initializing. I pushed [commit](https://github.com/ckeditor/ckeditor5-react/pull/480/commits/8d1cdcbe9a580e830e9844127d6dffcbf816f95e) that fixes the issue I think.

> The isLayoutReady prop in CKEditorContext works differently in the "classic" integration and multi-root integration. After providing isLayoutReady={false} in demo-multiroot-react-19, roots are rendered, and you can edit the content. However, when you pass the same prop in demo-react-19, the editor is not shown.

Is it correct behavior? It looks like our current implementation is broken (or it has another race condition). It makes absolutely no sense to initialize any editor within `ContextProvider` without waiting for initialization of context (because it will fall into the context adapter fallback in [these](https://github.com/ckeditor/ckeditor5-react/blob/ck/478/src/ckeditor.tsx#L225-L235) lines).

> Generally, I understand it was a good opportunity to refactor some code, but adding new functionalities made the overall changes unclear. I'm not sure which changes were necessary in the context of supporting React 19 and which were made for refactoring purposes. I believe we could have done it in two steps. Additionally, I need to rethink the additional types, guards, and so on.

Generally speaking it's complicated because it solves complex problem. React 19 tends to be more async than any other previous release and it only shown us hidden for many years issues with our implementation. The example might be `async` lifecycle methods in context class component which were reported already by community. These methods added other smaller race conditions. `StrictMode` another ones. `null` value in default context value too. This PR fixes all these smaller things to fix "big" Race Condition issue. 

> I believe we could have done it in two steps.

Agree however this PR is the first step. The next ones should be related to DX improvements and should utilize newer react hooks such as https://react.dev/reference/react/useSyncExternalStore which might help us with fixing with that kind of race condition issues (I hope so at least).

**Comment by Mati365** [comment]
Probably fixes this issue https://github.com/ckeditor/ckeditor5-react/issues/485#issue-2343726122. Must be checked.

**Comment by Mati365** [comment]
@Witoso  At this moment, nope. Too many changes. Initially, I thought that the watchdog race conditions were only reproducible on React 16-18. However, it turns out that they can also occur occasionally on older versions. On React 19, these race conditions happen consistently.

**Comment by Witoso** [comment]
Hmm, will it mean that context in next.js won't work until we release this PR?

**Comment by f1ames** [comment]
> @Witoso I suspect that it's reproducible on old installation methods, actually. Needs to be verified.

Yes, I checked with old installation methods and stable `ckeditor5-react` package and there is the same issue as reported in #485 for NIM setup.

**Comment by DawidKossowski** [comment]
> Generally speaking it's complicated because it solves complex problem.

@Mati365, sure, the problem is complex, and it is understandable. However, you have exported a lot of new things as a public API. Who should use it and when? We need to provide user guides describing the correct use cases and so on.

Moreover, we still want to refactor the whole integration. In that case, the public API could change. I would rather reduce the breaking changes in the future by only making internal changes that fix the issue instead of extending the API. That is what I meant.

**Comment by Mati365** [comment]
@DawidKossowski I thought it may be useful [here](https://github.com/ckeditor/ckeditor5-react/issues/409). I don't think that context format will be changed in future iterations so much. If you think we should not export it, I'll remove it.

**Comment by DawidKossowski** [comment]
That's what I meant by doing it in two steps. The first step will fix the integration with React 19 internally (the current PR), and the second step will enhance DX, the public API, and so on. I think the second step should be done as a separate task since it requires some research, analysis, and probably discussion. I'm not 100% sure about the current shape, and I don't want to block this PR, so I would remove it and address it in another task.

**Comment by Mati365** [comment]
@DawidKossowski gotcha, will be removed

**Review by DawidKossowski — APPROVED** [review]

**Comment by DawidKossowski** [comment]
I will merge it after releasing NIM.

**Comment by Witoso** [comment]
BTW, [React 19 is delayed.](https://www.cloudways.com/blog/react-v19-delay-suspense-tag-controversy/)

**Pull request merged**
